### PR TITLE
Milestone 1: Added Fusion G3 NN library with kernels related to add, mul, quantize…

### PR DESCRIPTION
**Pull request by ckmadhira**
…, dequantize, cat, layer norm, softmax to backends/cadence folder. Added operators to backends/cadence folder

### Summary
Added kernels and operators related to
1. Add
2. Mul,
3. Quantize
4. Dequatize
5. Cat
6. Layernorm
7. Softmax
to Fusion G3. The Kernels part of Fusion G3 NN library which is a submodule.

### CODE REVIEW

**Comment by ckmadhira** [comment]
Fusion G3 NN library is added with kernels related to add, mul, quantize, dequantize, cat, layernorm and softmax. Operators which use these kernels are also added to backends/cadence folder

**Review by zonglinpeng — CHANGES_REQUESTED** [review]
Thanks for the change! Had one comment on the op namespace

File: backends/cadence/aot/functions_fusion_g3.yaml [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+# aten ops
+- op: _to_copy.out
+  kernels:
+    - arg_meta: null
+      kernel_name: torch::executor::to_copy_out
+
+- op: _softmax.out
+  kernels:
+    - arg_meta: null
+      kernel_name: impl::FusionG3::softmax_out
```
[the remark this answers is not in the document]
Comment (zonglinpeng) [reply backends/cadence/aot/functions_fusion_g3.yaml:23]: Per @hsharma35 's comment above, please use cadence::impl::G3::native::<OP_NAME> namespace

File: backends/cadence/fusion_g3/operators/op_add.cpp [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+using exec_aten::Scalar;
+using exec_aten::ScalarType;
+using exec_aten::Tensor;
+using torch::executor::KernelRuntimeContext;
+using executorch::runtime::canCast;
+using torch::executor::Error;
+
+namespace impl {
+namespace FusionG3 { 
+namespace native {
```
[the remark this answers is not in the document]
Comment (ckmadhira) [reply backends/cadence/fusion_g3/operators/op_add.cpp:26]: Updated as per review comment

File: backends/cadence/aot/functions_fusion_g3.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ckmadhira) [reply backends/cadence/aot/functions_fusion_g3.yaml]: Updated the operator name as cadence::impl::G3::<OP_NAME>. native is not explicitly mentioned. As per kernel-library-custom-aten-kernel.md, native is automatically appended to the operators.

File: backends/cadence/aot/functions_fusion_g3.yaml [hunk 3]
```diff
@@ -0,0 +1,119 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
[the remark this answers is not in the document]
Comment (ckmadhira) [reply backends/cadence/aot/functions_fusion_g3.yaml:1]: Updated the operator name as cadence::impl::G3::add_out. native is not explicitly mentioned.

**Comment by zonglinpeng** [comment]
please make nnlib a standalone repository @ckmadhira

**Review by zonglinpeng — CHANGES_REQUESTED** [review]
comment as above

**Comment by ckmadhira** [comment]
> please make nnlib a standalone repository @ckmadhira

The decision is not to make the Fusion G3 NN library opensource. We are looking for a space where we can put the repo and provide exclusive access to Meta

**Comment by zonglinpeng** [comment]
@ckmadhira thanks for driving nnlib to a public repo! Can you remake this PR with just the kernel updates so that I can import as a clean change? Thank you
cc @mcremon-meta

**Comment by ckmadhira** [comment]
> @ckmadhira thanks for driving nnlib to a public repo! Can you remake this PR with just the kernel updates (i.e. everything except thirdparty nnlib) so that I can import as a clean change? Thank you cc @mcremon-meta

Updated Fusion G3 NN library as a submodule from opensource.

File: backends/cadence/aot/functions_fusion_g3.yaml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+- op: _softmax.out
+  kernels:
+    - arg_meta: null
+      kernel_name: cadence::impl::G3::softmax_out
+
+- op: add.out
+  kernels:
+    - arg_meta: null
+      kernel_name: cadence::impl::G3::add_out
```
Comment (hsharma35) [inline backends/cadence/aot/functions_fusion_g3.yaml:28]: @ckmadhira the name used here (`kernel_name: cadence::impl::G3::add_out`) does not match the actual kernel in `op_add.cpp` (`cadence::impl::G3::native::add_out`). Is this compiling on your end?

File: backends/cadence/aot/functions_fusion_g3.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (hsharma35) [reply backends/cadence/aot/functions_fusion_g3.yaml:1]: With the change to operator namespace, I think this would now fail to compile. Can you please check on your end?

**Review by hsharma35 — DISMISSED** [review]
LGTM, just needs one change:
Please make the operators in `op_*.cpp` consistent with the operators in `functions_fusion_g3.yaml`.

Thanks!

**Comment by zonglinpeng** [comment]
Please resolve the merge conflict in .gitmodules

File: backends/cadence/fusion_g3/third-party/nnlib/nnlib-FusionG3
Comment (zonglinpeng) [inline backends/cadence/fusion_g3/third-party/nnlib/nnlib-FusionG3:1]: Nice this is awesome!

File: backends/cadence/aot/functions_fusion_g3.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ckmadhira) [reply backends/cadence/aot/functions_fusion_g3.yaml:1]: We are able to compile and verify the operator with toy model. Are you having any issue with this name space?

**Comment by ckmadhira** [comment]
This is done

**Review by zonglinpeng — APPROVED** [review]

**Review by mcremon-meta — APPROVED** [review]

**Comment by mcremon-meta** [comment]
@ckmadhira there are a lot of linter errors, see https://github.com/pytorch/executorch/actions/runs/11950614633/job/33350989488?pr=6738. @zonglinpeng can you share instructions on how to run the linter?

**Comment by zonglinpeng** [comment]
> @ckmadhira there are a lot of linter errors, see https://github.com/pytorch/executorch/actions/runs/11950614633/job/33350989488?pr=6738. @zonglinpeng can you share instructions on how to run the linter?

https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#lintrunner @ckmadhira

**Review by zonglinpeng — CHANGES_REQUESTED** [review]
there are still some lint errors, please try `lintrunner -a`
https://github.com/pytorch/executorch/actions/runs/12009466002/job/33489114504?pr=6738

**Review by zonglinpeng — APPROVED** [review]
fixed lint on fb side

File: backends/cadence/aot/functions_fusion_g3.yaml [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+- op: where.self_out
+  kernels:
+    - arg_meta: null
+      kernel_name: torch::executor::where_out
+      
+- op: native_layer_norm.out
+  kernels:
+    - arg_meta: null
+      kernel_name: cadence::impl::G3::native_layer_norm_out
```
Comment (zonglinpeng) [inline backends/cadence/aot/functions_fusion_g3.yaml:118]: missing quant dequant per tensor func definition. fixed in  https://github.com/pytorch/executorch/pull/7061

File: backends/cadence/aot/functions_fusion_g3.yaml [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+# aten ops
+- op: _to_copy.out
+  kernels:
+    - arg_meta: null
+      kernel_name: torch::executor::to_copy_out
+
+- op: _softmax.out
+  kernels:
+    - arg_meta: null
+      kernel_name: cadence::impl::G3::softmax_out
```
Comment (zonglinpeng) [inline backends/cadence/aot/functions_fusion_g3.yaml:23]: need to be _softmax to be consistent, fixed in https://github.com/pytorch/executorch/pull/7061

File: backends/cadence/fusion_g3/operators/op_softmax.cpp [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+using exec_aten::Tensor;
+using torch::executor::Error;
+using torch::executor::KernelRuntimeContext;
+
+namespace cadence {
+namespace impl {
+namespace G3 { 
+namespace native {
+
+Tensor& softmax_out(
```
Comment (zonglinpeng) [inline backends/cadence/fusion_g3/operators/op_softmax.cpp:27]: same comment, _softmax,  fixed in https://github.com/pytorch/executorch/pull/7061

File: backends/cadence/fusion_g3/operators/op_quantize.cpp [hunk 8]
```diff
@@ -0,0 +1,797 @@
...
+ Bits4u = 21,
+ Bits4 = 22
+ };
+
+/**
+ * For an input tensor, use the scale and zero_point arguments to quantize it.
+ */
+namespace cadence {
+namespace impl {
+namespace FusionG3 {
```
Comment (zonglinpeng) [inline backends/cadence/fusion_g3/operators/op_quantize.cpp:38]: namespace needs to be G3.  fixed in https://github.com/pytorch/executorch/pull/7061

File: backends/cadence/fusion_g3/operators/op_dequantize.cpp [hunk 9]
```diff
@@ -0,0 +1,810 @@
...
+      "quant min: %" PRId64 " is greater than quant max: %" PRId64,
+      quant_min,
+      quant_max);
+}
+
+} // namespace
+
+
+/* Local function which calls the kernels based on the input datatype */
+void Dequantize_impl(Tensor& out,
```
Comment (zonglinpeng) [inline backends/cadence/fusion_g3/operators/op_dequantize.cpp:91]: please use all lower case  fixed in https://github.com/pytorch/executorch/pull/7061
